starlink/dish.py

```diff
@@ -8,430 +8,241 @@
 import logging
 import subprocess
 import threading
-
-import config
-
-from satellites import convert_observed, process_intervals
-
 from datetime import datetime, timezone, timedelta
 from pathlib import Path
-from config import DATA_DIR, STARLINK_GRPC_ADDR_PORT, DURATION_SECONDS, TLE_DATA_DIR
-from util import date_time_string, ensure_data_directory
-from obstruction import create_obstruction_map_video, process_obstruction_timeslot
+from typing import Optional, List, Dict, Any, Tuple
 
 import numpy as np
 import pandas as pd
 from skyfield.api import load
 
-logger = logging.getLogger(__name__)
+import config
+from config import DATA_DIR, STARLINK_GRPC_ADDR_PORT, DURATION_SECONDS, TLE_DATA_DIR
+from util import date_time_string, ensure_data_directory
+from obstruction import create_obstruction_map_video, process_obstruction_timeslot
+from data_processor import DataProcessor
+from satellite_estimator import SatelliteEstimator
+from grpc_command import GrpcCommand
+from dish_data_handler import DishDataHandler
+from timeslot_manager import TimeslotManager
 
+# Add starlink-grpc-tools to Python path
 sys.path.insert(0, str(Path("./starlink-grpc-tools").resolve()))
 import starlink_grpc
 
-GRPC_DATA_DIR = "{}/grpc".format(DATA_DIR)
+logger = logging.getLogger(__name__)
+
+# Constants for GRPC operations
+GRPC_DATA_DIR = f"{DATA_DIR}/grpc"
 GRPC_TIMEOUT = 10
 
 def grpc_get_location() -> None:
+    """Get and save dish location data."""
     name = "GRPC_GetLocation"
-    logger.info("{}, {}".format(name, threading.current_thread()))
-
-    FILENAME = "{}/{}/GetLocation-{}.txt".format(
-        GRPC_DATA_DIR, ensure_data_directory(GRPC_DATA_DIR), date_time_string()
-    )
-
-    # grpcurl -plaintext -d {\"get_location\":{}} 192.168.100.1:9200 SpaceX.API.Device.Device/Handle
-    cmd = [
-        "grpcurl",
-        "-plaintext",
-        "-d",
-        '{"get_location":{}}',
-        STARLINK_GRPC_ADDR_PORT,
-        "SpaceX.API.Device.Device/Handle",
-    ]
-    try:
-        with open(FILENAME, "w") as outfile:
-            subprocess.run(cmd, stdout=outfile, timeout=GRPC_TIMEOUT)
-    except subprocess.TimeoutExpired:
-        pass
+    logger.info(f"{name}, {threading.current_thread()}")
 
-    logger.info("Saved gRPC dish location to {}".format(FILENAME))
+    # Generate filename with current timestamp
+    filename = f"{GRPC_DATA_DIR}/{ensure_data_directory(GRPC_DATA_DIR)}/GetLocation-{date_time_string()}.txt"
     
+    # Create and execute GRPC command to get location
+    cmd = GrpcCommand("location", '{"get_location":{}}')
```

Comment:
instead of `get_location`, can you use `get_diagnostics` to get location data? `get_diagnostics` also returns `gpsTimeS`, which can then be converted to UTC time. 
In this way, the benefit is that the timing and location data will match more precisely instead of `time.time()` in our own logic

```python3
import gnss_lib_py as glp

gps_time = float(location["gpsTimeS"]) * 1e3
utc_time = glp.gps_to_unix_millis(gps_time)
```